presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatStreamReader.java

```diff
@@ -0,0 +1,429 @@
+    private final DateTimeZone hiveStorageTimeZone;
+    private final AggregatedMemoryContext systemMemoryContext;
+    // This is the StreamDescriptor for the value "column" with sequence ID 0, it is used to derive StreamDescriptors for the
+    // value "column" with other sequence IDs
+    private final StreamDescriptor baseValueStreamDescriptor;
+    private final OrcType.OrcTypeKind keyOrcType;
+
+    private final List<InputStreamSource<BooleanInputStream>> inMapStreamSources = new ArrayList<>();
+    private final List<BooleanInputStream> inMapStreams = new ArrayList<>();
+    private final List<StreamReader> valueStreamReaders = new ArrayList<>();
+    private final List<StreamDescriptor> valueStreamDescriptors = new ArrayList<>();
+
+    private Block dictionaryBlock;
+    private boolean[][] inMapVectors;
+    private int readOffset;
+    private int nextBatchSize;
+
+    @Nonnull
+    private InputStreamSource<BooleanInputStream> presentStreamSource = missingStreamSource(BooleanInputStream.class);
+    @Nullable
+    private BooleanInputStream presentStream;
+
+    private boolean rowGroupOpen;
+
+    public MapFlatStreamReader(StreamDescriptor streamDescriptor, DateTimeZone hiveStorageTimeZone, AggregatedMemoryContext systemMemoryContext)
+    {
+        this.streamDescriptor = requireNonNull(streamDescriptor, "stream is null");
+        this.hiveStorageTimeZone = requireNonNull(hiveStorageTimeZone, "hiveStorageTimeZone is null");
+        this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+        this.keyOrcType = streamDescriptor.getNestedStreams().get(0).getStreamType();
+        this.baseValueStreamDescriptor = streamDescriptor.getNestedStreams().get(1);
+    }
+
+    @Override
+    public void prepareNextRead(int batchSize)
+    {
+        readOffset += nextBatchSize;
+        nextBatchSize = batchSize;
+    }
+
+    @Override
+    public Block readBlock(Type type)
+            throws IOException
+    {
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        if (readOffset > 0) {
+            if (presentStream != null) {
+                // skip ahead the present bit reader, but count the set bits
+                // and use this as the skip size for the data reader
+                readOffset = presentStream.countBitsSet(readOffset);
+            }
+            if (readOffset > 0) {
+                for (int i = 0; i < valueStreamReaders.size(); i++) {
+                    int valueReadOffset = inMapStreams.get(i).countBitsSet(readOffset);
+                    valueStreamReaders.get(i).prepareNextRead(valueReadOffset);
+                }
+            }
+        }
+
+        boolean[] nullVector = null;
+
+        if (presentStream == null) {
+            for (int i = 0; i < inMapStreams.size(); i++) {
+                if (inMapStreams.get(i) != null) {
+                    inMapVectors[i] = new boolean[nextBatchSize];
```

Comment:
Do we want to set `inMapVectors[i]` to be `null if otherwise?  ditto for line 151

Reply:
> Do we want to set inMapVectors[i] to be null if otherwise? ditto for line 151

It looks like this is necessary in both cases for correctness. Improve test coverage?

Reply:
Technically this isn't even supported yet in DWRF, it will always provide all inMap streams.  I'll go back to the old way of allocating the array all at once (which is what would happen here eventually anyway), given we can't test it.

Reply:
It can always be added back once DWRF does support it.

Reply:
I remember, this was to avoid allocating the outer array repeatedly, I'll remove the if check since it's not necessary and may be confusing, and change it to an assertion